File: Behavioral/Command/DIY.java
package Command;

import java.util.ArrayDeque;
import java.util.Deque;

public class DIY {
    public static void main(String[] args) {
        var history = new History();
        var document = new WordProcessor();
        document.setContent("Hi!");

        var bold = new MakeBold(document, history);
        bold.execute();
        System.out.println(document.getContent());

        var italic = new MakeItalic(document, history);
        italic.execute();
        System.out.println(document.getContent());

        var underline = new MakeUnderline(document, history);
        underline.execute();
        System.out.println(document.getContent());

        var undo = new UndoCommand(history);
        undo.execute();
        undo.execute();
        undo.execute();
        System.out.println(document.getContent());

      
    }
}

class WordProcessor {
    private String content;

    public void makeItalic() {
        content = "<i>" + content + "</i>";
    }

    public void makeBold() {
        content = "<b>" + content + "</b>";
    }

    public void makeUnderline() {
        content = "<u>" + content + "</u>";
    }

    public void setContent(String content) {
        this.content = content;
    }

    public String getContent() {
        return content;
    }
}

class MakeItalic implements UndoableCommand {
    private String prevContent;
    private WordProcessor document;
    private History history;

    public MakeItalic(WordProcessor document, History history) {
        this.document = document;
        this.history = history;
    }

    @Override
    public void execute() {
        prevContent = document.getContent();
        document.makeItalic();
        history.push(this);
    }

    @Override
    public void unexecute() {
        document.setContent(prevContent);
    }
}

class MakeBold implements UndoableCommand {
    private String prevContent;
    private WordProcessor document;
    private History history;

    public MakeBold(WordProcessor document, History history) {
        this.document = document;
        this.history = history;
    }

    @Override
    public void execute() {
        prevContent = document.getContent();
        document.makeBold();
        history.push(this);
    }

    @Override
    public void unexecute() {
        document.setContent(prevContent);
    }
}

class MakeUnderline implements UndoableCommand {
    private String prevContent;
    private WordProcessor document;
    private History history;

    public MakeUnderline(WordProcessor document, History history) {
        this.document = document;
        this.history = history;
    }

    @Override
    public void execute() {
        prevContent = document.getContent();
        document.makeUnderline();
        history.push(this);
    }

    @Override
    public void unexecute() {
        document.setContent(prevContent);
    }
}

class UndoCommand implements Command {
    private History history;

    public UndoCommand(History history) {
        this.history = history;
    }

    @Override
    public void execute() {
        if (history.size() > 0) {
            history.pop().unexecute();
        }
    }
}

class History {
    Deque<UndoableCommand> commands = new ArrayDeque<>();

    public void push(UndoableCommand command) {
        commands.add(command);
    }

    public UndoableCommand pop() {
        return commands.pop();
    }

    public int size() {
        return commands.size();
    }
}

interface Command {
    void execute();
}

interface UndoableCommand extends Command {
    void unexecute();
}

File: TS/Behavioral/Command/DIY.ts
interface Command {
  execute(): void;
}

interface UndoableCommand extends Command {
  unexecute(): void;
}

class WordProcessor {
  private content: string;

  constructor() {
    this.content = "";
  }

  makeItalic(): void {
    this.content = `<i>${this.content}</i>`;
  }

  makeBold(): void {
    this.content = `<b>${this.content}</b>`;
  }

  makeUnderline(): void {
    this.content = `<u>${this.content}</u>`;
  }

  setContent(content: string): void {
    this.content = content;
  }

  getContent(): string {
    return this.content;
  }
}

class MakeItalic implements UndoableCommand {
  private prevContent: string;
  private document: WordProcessor;
  private commandhistory: CommandHistory;

  constructor(document: WordProcessor, commandhistory: CommandHistory) {
    this.document = document;
    this.commandhistory = commandhistory;
  }

  execute(): void {
    this.prevContent = this.document.getContent();
    this.document.makeItalic();
    this.commandhistory.push(this);
  }

  unexecute(): void {
    this.document.setContent(this.prevContent);
  }
}

class MakeBold implements UndoableCommand {
  private prevContent: string;
  private document: WordProcessor;
  private commandhistory: CommandHistory;

  constructor(document: WordProcessor, commandhistory: CommandHistory) {
    this.document = document;
    this.commandhistory = commandhistory;
  }

  execute(): void {
    this.prevContent = this.document.getContent();
    this.document.makeBold();
    this.commandhistory.push(this);
  }

  unexecute(): void {
    this.document.setContent(this.prevContent);
  }
}

class MakeUnderline implements UndoableCommand {
  private prevContent: string;
  private document: WordProcessor;
  private commandhistory: CommandHistory;

  constructor(document: WordProcessor, commandhistory: CommandHistory) {
    this.document = document;
    this.commandhistory = commandhistory;
  }

  execute(): void {
    this.prevContent = this.document.getContent();
    this.document.makeUnderline();
    this.commandhistory.push(this);
  }

  unexecute(): void {
    this.document.setContent(this.prevContent);
  }
}

class UndoCommand implements Command {
  private commandhistory: CommandHistory;

  constructor(commandhistory: CommandHistory) {
    this.commandhistory = commandhistory;
  }

  execute(): void {
    if (this.commandhistory.size() > 0) {
      const command = this.commandhistory.pop();
      if (command) {
        command.unexecute();
      }
    }
  }
}

class CommandHistory {
  private commands: UndoableCommand[] = [];
  // Deque<UndoableCommand> commands = new ArrayDeque<>();
  // const deque = new Deque<UndoableCommand>();

  push(command: UndoableCommand): void {
    this.commands.push(command);
  }

  pop(): UndoableCommand | undefined {
    return this.commands.pop();
  }

  size(): number {
    return this.commands.length;
  }
}

// The Client
const commandhistory = new CommandHistory();
const wordProcessor = new WordProcessor();
wordProcessor.setContent("Hi!");

const bold = new MakeBold(wordProcessor, commandhistory);
bold.execute();
console.log(wordProcessor.getContent());

const italic = new MakeItalic(wordProcessor, commandhistory);
italic.execute();
console.log(wordProcessor.getContent());

const underline = new MakeUnderline(wordProcessor, commandhistory);
underline.execute();
console.log(wordProcessor.getContent());

const undo = new UndoCommand(commandhistory);
undo.execute();
undo.execute();
undo.execute();
console.log(wordProcessor.getContent());
